### DISCUSSION TITLE
How to deal with empty cells?

### DISCUSSION THREAD

**Discussion by reckoner** [scikit-hep/boost-histogram#668] [Q&A] [answered]
I have managed to create a large 13 dimensional `Histogram`, but many of the entries are zero. I would like to identify the 13 dimensional coordinates of the zeros, but I don't understand how to do that? Most of the data in the Histogram is zero, and potentially even entire dimensions are all zero, so how would I discover that? 

Best regards,

**Comment by henryiii** [comment]
You can use the same methods you'd use on an array; `.values()` will give you the array, like `numpy.where`. Though if you have many zeroes, the indices list will be huge. I'm not sure what "entire dimensions will be zero" means - if one dimension is zero, the whole histogram is zero, unless you are using the flow bins and by zero you mean the hit was in a flow bin?

**Reply by reckoner** [reply]
Thank you! That helps. I have been reading the documentation but I don't understand what a *flow bin* is.

**Reply by henryiii** [reply]
Let's say you have a histogram with two axes:

```python
>>> import boost_histogram as bh
>>> h = bh.Histogram(bh.axis.Regular(10, 0, 10), bh.axis.Regular(10, -1, 1))
```

Now, if we fill it with a hit at (3, 3), this will be valid in the first axes (0 to 10), but will "miss" the second axis (-1 to 1). However, as long as you have flow bins on for that axis (the default), it will get registered in the overflow bin.

```python
>>> h.fill(3, 3)
Histogram(
  Regular(10, 0, 10),
  Regular(10, -1, 1),
  storage=Double()) # Sum: 0.0 (1.0 with flow)
```

Why is this useful (other than having both a in-range sum `h.sum()` and a with-flow sum `h.sum(flow=True)`, both of which show up in the default repr)? This is very useful if you project:

```python
>>> h.project(0)
Histogram(Regular(10, 0, 10), storage=Double()) # Sum: 1.0
```

Or, equivalently, summing over an axis:
```python
>>> h[:, sum]
Histogram(Regular(10, 0, 10), storage=Double()) # Sum: 1.0
```

This is identical to if you had filled a histogram with just the first axis, ignoring the limits on the second axis entirely! Often this saves time, you don't need multiple histograms if you don't want to have one limit affect another axis.

If you don't want the flow bins in the sum, then place limits when you sum:

```python
>>> h[:, 0:len:sum]
Histogram(Regular(10, 0, 10), storage=Double())
```

**Reply by reckoner** [reply]
Very clear explanation! I have 13 axes. How would I project down to the 8th and 10th axes (jointly), for example?

**Reply by henryiii** [reply]
`.project(7, 9)`. If you use Hist, you can add names then use names here (I'd love to add that to bh too, but don't have permission to do so, and may never have).

**Reply by reckoner** [reply]
what is `Hist` and how is it different from `bh`?

**Reply by henryiii** [reply]
Hist is an extension of boost-histogram with extra methods and functionality to make it easier to use, especially in a REPL or notebook. `name=` is the one thing I consider to be a core feature that most code should use that boost-histogram is missing, but the various shortcuts are also nice, and support for `.plot()` on a histogram comes in handy. It also has a few extra features, like support for stacks of histograms.

**Comment by reckoner** [comment, marked as answer]
OK. Here the [Hist link ](https://github.com/scikit-hep/hist) for posterity.

**Reply by henryiii** [reply]
It's at the top of boost-histogram's readme, too. :)
